{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# Background"
      ],
      "metadata": {
        "id": "kWrBEjNqng49"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "This notebook contains all necessary steps for pre-processing and variant calling analysis. The objective for this notebook is to conduct variant calling analysis for a single gene of interest in both pre-treatment and during/post-treatment samples of a single patient in the study described in *Circulating tumor DNA sequencing in colorectal cancer patients treated with first-line chemotherapy with anti-EGFR*. The raw FASTQ files for this study are available on NCBI under BioProject PRJNA714799."
      ],
      "metadata": {
        "id": "w42GMBRjkDLM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Excercise Specific Information"
      ],
      "metadata": {
        "id": "Q7rgjTmNnqSO"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Patient: CTC360\n",
        "\n",
        "Pre-Treatment Sample Reads:\n",
        "- Biosample - https://www.ncbi.nlm.nih.gov/biosample/18318230\n",
        "- SRA - https://www.ncbi.nlm.nih.gov/sra?LinkName=biosample_sra&from_uid=18318230\n",
        "- SRR - https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR13973704&display=metadata\n",
        "- FASTQ - SRR13973704.fastq\n",
        "\n",
        "During/Post-Treatment Sample Reads:\n",
        "- Biosample -  https://www.ncbi.nlm.nih.gov/biosample/18318229\n",
        "- SRA - https://www.ncbi.nlm.nih.gov/sra?LinkName=biosample_sra&from_uid=18318229\n",
        "- SRR - https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR13973705&display=metadata\n",
        "- FASTQ - SRR13973705.fastq\n",
        "\n",
        "Reference Genome: hg38\n",
        "- https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001405.26/\n",
        "\n",
        "Known Variants VCF: dbSNP (latest release)\n",
        "- https://ftp.ncbi.nih.gov/snp/latest_release/VCF/\n",
        "\n",
        "Gene of Interest: TP53\n",
        "- Refseq - https://www.ncbi.nlm.nih.gov/gene/7157\n",
        "- UCSC GB - https://genome.ucsc.edu/cgi-bin/hgGene?hgg_gene=ENST00000618944.4&hgg_chrom=chr17&hgg_start=7668401&hgg_end=7675493&hgg_type=knownGene&db=hg38\n",
        "- Acts as a tumor suppressor in many tumor types; induces growth arrest or apoptosis depending on the physiological circumstances and cell type. Involved in cell cycle regulation as a trans-activator that acts to negatively regulate cell division by controlling a set of genes required for this process.\n",
        "- 71 mutations over 64 patients (68.8%) at baseline in study.\n",
        "- The TP53 gene is the most frequently mutated gene (>50%) in human cancer, indicating that the TP53 gene plays a crucial role in preventing cancer formation.\n"
      ],
      "metadata": {
        "id": "2zg_q1yhn4W5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Pre-Processing"
      ],
      "metadata": {
        "id": "Db9XIOs5oflF"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Pre-processing of sequence reads in FASTQ files downloaded from SRA was conducted in adherence with *Best practices for variant calling in clinical sequencing* (https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00791-w) and GATK's best practice workflows for data pre-processing for variant discovery (https://gatk.broadinstitute.org/hc/en-us/articles/360035535912-Data-pre-processing-for-variant-discovery)."
      ],
      "metadata": {
        "id": "AD88A-gZos9i"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 1: Generate index file for hg38 reference genome.**\n",
        "- This step is necessary for alignment using BWA MEM.\n",
        "- hg38 was used, instead of hg19 (used in study), since it contains the most comprehensive and up to date representation of the human reference genome.\n"
      ],
      "metadata": {
        "id": "eK-9nIRjpz-Y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "bwa index -p hg38 data/hg38_ref.fa"
      ],
      "metadata": {
        "id": "kBsJ6n6XqVmG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 2: Align sequence reads to hg38 reference genome.**"
      ],
      "metadata": {
        "id": "BOzyyQSorTku"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "bwa mem hg38 data/SRR13973704.fastq > data/SRR13973704.sam\n",
        "\n",
        "# During/post-treatment\n",
        "bwa mem hg38 data/SRR13973705.fastq > data/SRR13973705.sam"
      ],
      "metadata": {
        "id": "dlfMIbgUr7qq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 3: Convert SAM to BAM**\n",
        "- Alignment step outputs SAM file.\n",
        "- Variant calling analysis using GATK requires BAM file as input.\n",
        "- S option specifies input file is in SAM format.\n",
        "- b option specifies output file is in BAM format."
      ],
      "metadata": {
        "id": "xODQTo6isJq7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "samtools view -Sb data/SRR13973704.sam > data/SRR13973704.bam\n",
        "\n",
        "# During/post-treatment\n",
        "samtools view -Sb data/SRR13973705.sam > data/SRR13973705.bam"
      ],
      "metadata": {
        "id": "MzyJeIbsslaO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 4: Sort BAM files**"
      ],
      "metadata": {
        "id": "FfkEsEEHugAm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "samtools sort data/SRR13973704.bam > data/SRR13973704_sorted.bam\n",
        "\n",
        "# During/post-treatment\n",
        "samtools sort data/SRR13973705.bam > data/SRR13973705_sorted.bam"
      ],
      "metadata": {
        "id": "ol-OaHaXuqra"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 5: Index sorted BAM files**\n",
        "- Necessary for filtering BAM by region.\n",
        "- BAM filtering by region necessary to reduce uneccessary processing."
      ],
      "metadata": {
        "id": "GTlzd_Gy1r1p"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "samtools index data/SRR13973704_sorted.bam\n",
        "\n",
        "# During/Post-treatment\n",
        "samtools index data/SRR13973705_sorted.bam\n"
      ],
      "metadata": {
        "id": "S0jWSTcE2ceQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 6: Filter BAM by region**\n",
        "- Only interested in TP53\n",
        "- TP53 refseq coordinates - NC_000017.11:7668421-7687490"
      ],
      "metadata": {
        "id": "0h0xrohY2z6Y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "samtools view -b data/SRR13973704_sorted.bam \"NC_000017.11:7668421-7687490\" > data/SRR13973704_tp53.bam\n",
        "\n",
        "# During/post-treatment\n",
        "samtools view -b data/SRR13973705_sorted.bam \"NC_000017.11:7668421-7687490\" > data/SRR13973705_tp53.bam"
      ],
      "metadata": {
        "id": "n9ccr3dE3Jxs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 7: Add read groups**\n",
        "- Necessary since read group information not included in FASTQ.\n",
        "- Read group headers are required for base quality score recalibration.\n",
        "- Relevant information available in SRR."
      ],
      "metadata": {
        "id": "WVHOiyUZ4gO-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "gatk AddOrReplaceReadGroups -I data/SRR13973704_tp53.bam -O data/SRR13973704_tp53_RG.bam -RGID SRR13973704 -RGLB plasma_266 -RGPL ILLUMINA -RGSM SRR13973704 -RGPU SRR13973704\n",
        "\n",
        "# During/post-treatment\n",
        "gatk AddOrReplaceReadGroups -I data/SRR13973705_tp53.bam -O data/SRR13973705_tp53_RG.bam -RGID SRR13973705 -RGLB plasma_266 -RGPL ILLUMINA -RGSM SRR13973705 -RGPU SRR13973705\n"
      ],
      "metadata": {
        "id": "eGV-6lNT38Wl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 8: Generate reference FASTA index file**\n",
        "- Necessary for base quality score recalibration."
      ],
      "metadata": {
        "id": "iNq15a63DoK-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "samtools faidx data/hg38_ref.fa"
      ],
      "metadata": {
        "id": "fdTmdHRkEM-Z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 9: Create reference sequence dictionary**\n",
        "- Necessary for base quality score recalibration."
      ],
      "metadata": {
        "id": "SwzSymtcER-u"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "gatk CreateSequenceDictionary -R data/hg38_ref.fa -O data/hg38_ref.dict"
      ],
      "metadata": {
        "id": "_UfHdC_OEbIR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 10: Index known sites VCF file**\n",
        "- Necessary for base quality score recalibration."
      ],
      "metadata": {
        "id": "MRzroY6AExd7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "gatk IndexFeatureFile -I data/GCF_000001405_40.vcf"
      ],
      "metadata": {
        "id": "3rPBDuZxFbz0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 11: Generate recalibration table**\n",
        "- Necessary for base quality score recalibration."
      ],
      "metadata": {
        "id": "dn-XZlfFF1Q0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "gatk BaseRecalibrator -I data/SRR13973704_tp53_RG.bam -R data/hg38_ref.fa --known-sites data/GCF_000001405_40.vcf -O data/SRR13973704_tp53_recal_data.table\n",
        "\n",
        "# During/post-treatment\n",
        "gatk BaseRecalibrator -I data/SRR13973705_tp53_RG.bam -R data/hg38_ref.fa --known-sites data/GCF_000001405_40.vcf -O data/SRR13973705_tp53_recal_data.table\n"
      ],
      "metadata": {
        "id": "zuk1l1UBF2n1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 12: Base quality score recalibration**\n",
        "- Detects systematic errors made by the sequencing machine when it estimates the accuracy of each base call.\n",
        "- Short variant calling algorithms rely heavily on the quality score assigned to the individual base calls in each sequence read.\n",
        "- Scores produced by the machines are subject to various sources of systematic (non-random) technical error, leading to over- or under-estimated base quality scores in the data."
      ],
      "metadata": {
        "id": "Rgj1jr7QG43k"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "gatk ApplyBQSR -R data/hg38_ref.fa -I data/SRR13973704_tp53_RG.bam --bqsr-recal-file data/SRR13973704_tp53_recal_data.table -O data/SRR13973704_tp53_recal.bam\n",
        "\n",
        "# During/post-treatment\n",
        "gatk ApplyBQSR -R data/hg38_ref.fa -I data/SRR13973705_tp53_RG.bam --bqsr-recal-file data/SRR13973705_tp53_recal_data.table -O data/SRR13973705_tp53_recal.bam\n"
      ],
      "metadata": {
        "id": "DTCTaaG-HCkB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Variant Calling Analysis\n",
        "Variant calling analysis was conducted in adherence with GATK's best practice workflow for somatic short variant discovery (SNVs + Indels) (https://gatk.broadinstitute.org/hc/en-us/articles/360035894731-Somatic-short-variant-discovery-SNVs-Indels)."
      ],
      "metadata": {
        "id": "dL39qi4pHrgj"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 13: Variant calling**\n",
        "- GATK Mutect2 used as it is HaplotypeCaller equivalent for somatic cells."
      ],
      "metadata": {
        "id": "NHqJfFsVqea6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "gatk Mutect2 -R data/hg38_ref.fa -I data/SRR13973704_tp53_recal.bam -O data/SRR13973704.vcf.gz\n",
        "\n",
        "# During/post-treatment\n",
        "gatk Mutect2 -R data/hg38_ref.fa -I data/SRR13973705_tp53_recal.bam -O data/SRR13973705.vcf.gz\n"
      ],
      "metadata": {
        "id": "kAg-HzrAHwi7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 14: Filter variants**"
      ],
      "metadata": {
        "id": "ylsTjd-HqwyV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "gatk FilterMutectCalls -R data/hg38_ref.fa -V data/SRR13973704.vcf.gz -O data/SRR13973704_filtered.vcf.gz\n",
        "\n",
        "#During/post-treatment\n",
        "gatk FilterMutectCalls -R data/hg38_ref.fa -V data/SRR13973705.vcf.gz -O data/SRR13973705_filtered.vcf.gz\n"
      ],
      "metadata": {
        "id": "TuwQk_g9q9-K"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 15: Index ClinVar Annotation VCF**\n",
        "- Necessary for SnpSift"
      ],
      "metadata": {
        "id": "3mcRXOclWeZX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "tabix -p vcf data/clinvar_20251109.vcf.gz"
      ],
      "metadata": {
        "id": "wstyozVSWrgO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 16: Rename Chr in variants VCF**\n",
        "- Necessary to ensure proper mapping between ClinVar and variants for annotation.\n",
        "- ClinVar uses integer notation to represent chomosomes.\n",
        "- SRR uses RefSeq notation to represent loci."
      ],
      "metadata": {
        "id": "Moz9Ev3zkKmW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "bcftools annotate --rename-chrs chr_name_conv.txt data/SRR13973704_filtered.vcf.gz -o data/SRR13973704_filtered_renamed.vcf\n",
        "bgzip data/SRR13973704_filtered_renamed.vcf\n",
        "\n",
        "# During/post-treatment\n",
        "bcftools annotate --rename-chrs chr_name_conv.txt data/SRR13973705_filtered.vcf.gz -o data/SRR13973705_filtered_renamed.vcf\n",
        "bgzip data/SRR13973705_filtered_renamed.vcf"
      ],
      "metadata": {
        "id": "HW_zSZjCkaSa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 15: Annotate Variants**\n",
        "- Annotations from ClinVar databases"
      ],
      "metadata": {
        "id": "rhx7YgLlGYuq"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%bash\n",
        "# Pre-treatment\n",
        "SnpSift annotate data/clinvar_20251109.vcf.gz data/SRR13973704_filtered_renamed.vcf.gz > data/SRR13973704_annotated.vcf\n",
        "\n",
        "# During/post-treatment\n",
        "SnpSift annotate data/clinvar_20251109.vcf.gz data/SRR13973705_filtered_renamed.vcf.gz > data/SRR13973705_annotated.vcf"
      ],
      "metadata": {
        "id": "WaEwYAitGJf1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Step 16: Identify Clinically Relevant Variants**\n",
        "- Variants considered clinically relevant if found in ClinVar."
      ],
      "metadata": {
        "id": "kw5gIXBiJOgl"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "# Import VCF files to DataFrame\n",
        "\n",
        "pre_treatment_variants = pd.read_csv('data/SRR13973704_annotated.vcf', comment='#', sep='\\t',\n",
        "                                     names=['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SRR13973704'])\n",
        "post_treatment_variants = pd.read_csv('data/SRR13973705_annotated.vcf', comment='#', sep='\\t',\n",
        "                                      names=['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'SRR13973704'])\n",
        "\n",
        "# Filter variants based w/ ClinVar Annotation\n",
        "clinically_relevant_pre_treatment_variants = pre_treatment_variants[pre_treatment_variants['ID'] != '.']\n",
        "clinically_relevant_post_treatment_variants = post_treatment_variants[post_treatment_variants['ID'] != '.']\n",
        "\n",
        "# Display clinically relevant mutations\n",
        "clinically_relevant_pre_treatment_variants\n",
        "clinically_relevant_post_treatment_variants\n"
      ],
      "metadata": {
        "id": "NXel6ZPFJnD0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Discussion\n",
        "\n",
        "The goal of this exercise was to conduct variant calling analysis (VCA) for a specific gene of interest on both pre-treatment and during/post-treatment samples from of a singular patient . Specifically, I conducted VCA for TP53 on both samples from patient CTC360. I was able to identify a decrease in both the total number of variants (79 -> 45), and the number clinically relevant variants (16 -> 11), following chemotherapy treatment with cetuximab-containing regimen. This is consistent with findings from *Circulating tumor DNA sequencing in colorectal cancer patients treated with first-line chemotherapy with anti-EGFR*, which showed significant decreases in Variant Allele Frequency (VAF) following treatment in most patients. Additionally, *Lim et al.* showed that the change of average VAF between the baseline and the first response evaluation had a linear correlation with tumor size changes on CT images. TP53 acts as a tumor suppressor in many tumor types -- inducing growth arrest or apoptosis depending on the physiological circumstances and cell type -- and is involved in cell cycle regulation as a trans-activator that acts to negatively regulate cell division by controlling a set of genes required for this process.Since TP53 is known to confer tumor progression, this result could indicate that response to treatment may be driven by reduced variation in the TP53 region. This said, given the data considered in this exercise and the sample size, it is impossible to prove this hypothesis.\n",
        "\n",
        "\n",
        "When displaying these findings for a presentation, I would display the variant information tables from VCF in a tabular format. Additionally, to clearly show the change in number of variants, I would include a bar plot with counts for total variants and clinically relevant variants, pre and during/post-treatment. I would also do a deeper dive into specific variants of interest, including relevant information from ClinVar and COSMIC. I would also include some graphics showing an overview of the EGFR pathway to explain why alterations in this pathwayh could lead to disease progression."
      ],
      "metadata": {
        "id": "drhBIFncsXZr"
      }
    }
  ]
}